From 4295a6ee3675c98499f59b82e7195933b9d8de43 Mon Sep 17 00:00:00 2001
From: Gowtham Anandha Babu <gowtham.anandha.babu@intel.com>
Date: Mon, 5 Aug 2024 16:50:34 +0000
Subject: [PATCH] Modify BT aidl hal for bluetooth bring up

BT fails to turn on with AOSP aidl hal as it is.

Modify the power on/off sequence of BT controller to bring
the Bluetooth up on aidl hal.

Tests done:
1. Flash binary
2. Check BT on/off
3. Check BT scan
4. Check BT connect to headset

Tracked-On: OAM-122186
Signed-off-by: Gowtham Anandha Babu <gowtham.anandha.babu@intel.com>
Signed-off-by: Jeevaka Prabu Badrappan <jeevaka.badrappan@intel.com>
---
 bluetooth/aidl/default/Android.bp             | 28 +++++++++----------
 bluetooth/aidl/default/BluetoothHci.cpp       |  2 +-
 ...lt.rc => bluetooth-service-default.vbt.rc} |  2 +-
 ....xml => bluetooth-service-default.vbt.xml} |  0
 bluetooth/aidl/default/net_bluetooth_mgmt.cpp | 19 +++++++++----
 bluetooth/aidl/default/service.cpp            |  2 +-
 bluetooth/hci/Android.bp                      |  8 +++---
 7 files changed, 35 insertions(+), 26 deletions(-)
 rename bluetooth/aidl/default/{bluetooth-service-default.rc => bluetooth-service-default.vbt.rc} (86%)
 rename bluetooth/aidl/default/{bluetooth-service-default.xml => bluetooth-service-default.vbt.xml} (100%)

diff --git a/bluetooth/aidl/default/Android.bp b/bluetooth/aidl/default/Android.bp
index 3f4ba99..9f0f5b4 100644
--- a/bluetooth/aidl/default/Android.bp
+++ b/bluetooth/aidl/default/Android.bp
@@ -3,7 +3,7 @@ package {
 }
 
 cc_defaults {
-    name: "android.hardware.bluetooth-service-build-defaults",
+    name: "android.hardware.bluetooth-service-build-defaults-vbt",
     cflags: [
         "-Wall",
         "-Wextra",
@@ -19,15 +19,15 @@ cc_defaults {
     ],
     static_libs: [
         "android.hardware.bluetooth.async",
-        "android.hardware.bluetooth.hci",
+        "android.hardware.bluetooth.hci.vbt",
     ],
 }
 
 cc_library_static {
-    name: "libbluetoothhcihalimpl",
+    name: "libbluetoothhcihalimplvbt",
     vendor_available: true,
     host_supported: true,
-    defaults: ["android.hardware.bluetooth-service-build-defaults"],
+    defaults: ["android.hardware.bluetooth-service-build-defaults-vbt"],
     srcs: [
         "BluetoothHci.cpp",
         "net_bluetooth_mgmt.cpp",
@@ -35,12 +35,12 @@ cc_library_static {
 }
 
 cc_binary {
-    name: "android.hardware.bluetooth-service.default",
+    name: "android.hardware.bluetooth-service.default.vbt",
     relative_install_path: "hw",
-    init_rc: ["bluetooth-service-default.rc"],
-    vintf_fragments: [":manifest_android.hardware.bluetooth-service.default.xml"],
+    init_rc: ["bluetooth-service-default.vbt.rc"],
+    vintf_fragments: [":manifest_android.hardware.bluetooth-service.default.vbt.xml"],
     vendor: true,
-    defaults: ["android.hardware.bluetooth-service-build-defaults"],
+    defaults: ["android.hardware.bluetooth-service-build-defaults-vbt"],
     srcs: [
         "service.cpp",
     ],
@@ -53,12 +53,12 @@ cc_binary {
         "liblog",
     ],
     static_libs: [
-        "libbluetoothhcihalimpl",
+        "libbluetoothhcihalimplvbt",
     ],
 }
 
 cc_fuzz {
-    name: "android.hardware.bluetooth-service.default_fuzzer",
+    name: "android.hardware.bluetooth-service.default.vbt_fuzzer",
     host_supported: true,
     defaults: ["service_fuzzer_defaults"],
     srcs: [
@@ -66,9 +66,9 @@ cc_fuzz {
     ],
     static_libs: [
         "android.hardware.bluetooth.async",
-        "android.hardware.bluetooth.hci",
+        "android.hardware.bluetooth.hci.vbt",
         "android.hardware.bluetooth-V1-ndk",
-        "libbluetoothhcihalimpl",
+        "libbluetoothhcihalimplvbt",
         "liblog",
     ],
     fuzz_config: {
@@ -80,6 +80,6 @@ cc_fuzz {
 }
 
 filegroup {
-    name: "manifest_android.hardware.bluetooth-service.default.xml",
-    srcs: ["bluetooth-service-default.xml"],
+    name: "manifest_android.hardware.bluetooth-service.default.vbt.xml",
+    srcs: ["bluetooth-service-default.vbt.xml"],
 }
diff --git a/bluetooth/aidl/default/BluetoothHci.cpp b/bluetooth/aidl/default/BluetoothHci.cpp
index a247cb0..3f0774f 100644
--- a/bluetooth/aidl/default/BluetoothHci.cpp
+++ b/bluetooth/aidl/default/BluetoothHci.cpp
@@ -14,7 +14,7 @@
  * limitations under the License.
  */
 
-#define LOG_TAG "android.hardware.bluetooth.service.default"
+#define LOG_TAG "android.hardware.bluetooth.service.default.vbt"
 
 #include "BluetoothHci.h"
 
diff --git a/bluetooth/aidl/default/bluetooth-service-default.rc b/bluetooth/aidl/default/bluetooth-service-default.vbt.rc
similarity index 86%
rename from bluetooth/aidl/default/bluetooth-service-default.rc
rename to bluetooth/aidl/default/bluetooth-service-default.vbt.rc
index dc78698..4382fc5 100644
--- a/bluetooth/aidl/default/bluetooth-service-default.rc
+++ b/bluetooth/aidl/default/bluetooth-service-default.vbt.rc
@@ -1,4 +1,4 @@
-service vendor.bluetooth-default /vendor/bin/hw/android.hardware.bluetooth-service.default
+service vendor.bluetooth-default /vendor/bin/hw/android.hardware.bluetooth-service.default.vbt
     class hal
     capabilities BLOCK_SUSPEND NET_ADMIN SYS_NICE
     user bluetooth
diff --git a/bluetooth/aidl/default/bluetooth-service-default.xml b/bluetooth/aidl/default/bluetooth-service-default.vbt.xml
similarity index 100%
rename from bluetooth/aidl/default/bluetooth-service-default.xml
rename to bluetooth/aidl/default/bluetooth-service-default.vbt.xml
diff --git a/bluetooth/aidl/default/net_bluetooth_mgmt.cpp b/bluetooth/aidl/default/net_bluetooth_mgmt.cpp
index 0699781..664dc74 100644
--- a/bluetooth/aidl/default/net_bluetooth_mgmt.cpp
+++ b/bluetooth/aidl/default/net_bluetooth_mgmt.cpp
@@ -14,7 +14,7 @@
  * limitations under the License.
  */
 
-#define LOG_TAG "android.hardware.bluetooth.service.default"
+#define LOG_TAG "android.hardware.bluetooth.service.default.vbt"
 
 #include "net_bluetooth_mgmt.h"
 
@@ -23,6 +23,7 @@
 #include <poll.h>
 #include <sys/socket.h>
 #include <unistd.h>
+#include <sys/ioctl.h>
 
 #include <cerrno>
 #include <cstdint>
@@ -74,6 +75,8 @@ struct mgmt_ev_read_index_list {
 #define RFKILL_OP_ADD 0
 #define RFKILL_OP_CHANGE 2
 
+#define IOCTL_HCIDEVDOWN _IOW('H', 202, int)
+
 struct rfkill_event {
   uint32_t idx;
   uint8_t type;
@@ -249,8 +252,8 @@ int NetBluetoothMgmt::rfkill(int block) {
 int NetBluetoothMgmt::openHci(int hci_interface) {
   ALOGI("opening hci interface %d", hci_interface);
 
-  // Block Bluetooth.
-  rfkill(1);
+  // Unblock Bluetooth.
+  rfkill(0);
 
   // Wait for the HCI interface to complete initialization or to come online.
   hci_interface = waitHciDev(hci_interface);
@@ -272,6 +275,12 @@ int NetBluetoothMgmt::openHci(int hci_interface) {
       .hci_channel = HCI_CHANNEL_USER,
   };
 
+  /* Force interface down to use HCI user channel */
+  if (ioctl(fd, IOCTL_HCIDEVDOWN, hci_interface)) {
+    ALOGE("HCIDEVDOWN ioctl error: %s", strerror(errno));
+    return -1;
+  }
+
   // Bind the socket to the selected interface.
   if (bind(fd, (struct sockaddr*)&hci_addr, sizeof(hci_addr)) < 0) {
     ALOGE("unable to bind bluetooth user channel: %s", strerror(errno));
@@ -290,8 +299,8 @@ void NetBluetoothMgmt::closeHci() {
     bt_fd_ = -1;
   }
 
-  // Unblock Bluetooth.
-  rfkill(0);
+  // Block Bluetooth.
+  rfkill(1);
 }
 
 }  // namespace aidl::android::hardware::bluetooth::impl
diff --git a/bluetooth/aidl/default/service.cpp b/bluetooth/aidl/default/service.cpp
index ef4b884..69c60a9 100644
--- a/bluetooth/aidl/default/service.cpp
+++ b/bluetooth/aidl/default/service.cpp
@@ -14,7 +14,7 @@
  * limitations under the License.
  */
 
-#define LOG_TAG "aidl.android.hardware.bluetooth.service.default"
+#define LOG_TAG "aidl.android.hardware.bluetooth.service.default.vbt"
 
 #include <aidl/android/hardware/bluetooth/IBluetoothHci.h>
 #include <android/binder_manager.h>
diff --git a/bluetooth/hci/Android.bp b/bluetooth/hci/Android.bp
index f0f6e8f..72c2dd4 100644
--- a/bluetooth/hci/Android.bp
+++ b/bluetooth/hci/Android.bp
@@ -1,9 +1,9 @@
 package {
-    default_applicable_licenses: ["hardware_interfaces_license"],
+    default_applicable_licenses: ["Android-Apache-2.0"],
 }
 
 cc_library_static {
-    name: "android.hardware.bluetooth.hci",
+    name: "android.hardware.bluetooth.hci.vbt",
     vendor_available: true,
     host_supported: true,
     defaults: ["hidl_defaults"],
@@ -21,7 +21,7 @@ cc_library_static {
 }
 
 cc_test {
-    name: "bluetooth-vendor-interface-hci-test",
+    name: "bluetooth-vendor-interface-hci-vbt-test",
     host_supported: true,
     defaults: ["hidl_defaults"],
     srcs: [
@@ -35,7 +35,7 @@ cc_test {
     ],
     static_libs: [
         "android.hardware.bluetooth.async",
-        "android.hardware.bluetooth.hci",
+        "android.hardware.bluetooth.hci.vbt",
         "libgmock",
     ],
     sanitize: {
-- 
2.34.1

